### PR TITLE
Compressed

### PR DESCRIPTION
**Pull request by jannikluhn**
### What was wrong?

Missing support for compressed public keys (#55).

### How was it fixed?

- Add `to/from_compressed_bytes` methods to `PublicKey` class.
- Add implementations for the coincurve and native backend.
- Add some tests.

I'm not an expert in cryptography, so please review carefully, especially the native backend code.

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

File: eth_keys/backends/base.py [hunk 1]
```diff
@@ -27,3 +27,11 @@ def ecdsa_recover(self,
     def private_key_to_public_key(self,
                                   private_key: PrivateKey) -> PublicKey:
         raise NotImplementedError()
+
+    def decompress_public_key_bytes(self,
+                                    compressed_public_key_bytes: bytes) -> bytes:
+        raise NotImplementedError()
```
Comment (pipermerriam) [inline eth_keys/backends/base.py:33]: Note to self that we should convert these to be `ABC,abstractmethod` based sometime.

File: eth_keys/validation.py [hunk 2]
```diff
@@ -56,10 +56,20 @@ def validate_message_hash(value: Any) -> None:
...
+        raise ValidationError(
+            "Unexpected uncompressed public key format.  Must be length 64 byte string"
+        )
+
+
+def validate_compressed_public_key_bytes(value: Any) -> None:
+    validate_bytes(value)
+    if len(value) != 33:
+        raise ValidationError(
+            "Unexpected compressed public key format.  Must be length 33 byte string"
```
Comment (pipermerriam) [inline eth_keys/validation.py]: Nice to include the actual value length here as well so it's clear what the discrepancy is.

**Review by pipermerriam — COMMENTED** [review]
Looks like there is one more easy test that could be written to randomly generate a private key and validate that `decompress(compress(original)) == original` is always true (using hypothesis)

Also needs to update the documentation with these new APIs.

**Comment by jannikluhn** [comment]
Added the suggested hypothesis test, improved the validation error message, and updated the readme.

Also, when looking at the readme I realized that the key compression API is duplicated: Once in PublicKey (where one would naturally look for it) and once in `KeyAPI` (where it doesn't make too much sense IMO). I therefore removed it from `KeyAPI` and udpated the tests accordingly.

File: eth_keys/backends/native/ecdsa.py [hunk 3]
```diff
@@ -60,6 +60,35 @@ def private_key_to_public_key(private_key_bytes: bytes) -> bytes:
     return public_key_bytes
 
 
+def compress_public_key(uncompressed_public_key_bytes: bytes) -> bytes:
+    x, y = decode_public_key(uncompressed_public_key_bytes)
+    if y % 2 == 0:
+        prefix = b"\x02"
+    else:
+        prefix = b"\x03"
+    return prefix + pad32(int_to_big_endian(x))
```
Comment (carver) [inline eth_keys/backends/native/ecdsa.py:69]: lol, padding the compressed key is a strange spec. I was wondering why the compressed key is always the same length...

File: tests/backends/test_native_backend_against_coincurve.py [hunk 4]
```diff
@@ -157,3 +147,62 @@ def test_coincurve_to_native_invalid_signatures(message_hash,
     public_key_b = backend_b.ecdsa_recover(message_hash, signature_a)
 
     assert public_key_b == public_key_a
+
+
+@given(
+    private_key_bytes=private_key_st,
+    direction=st.one_of(
+        st.just('coincurve-to-native'),
```
Comment (carver) [inline tests/backends/test_native_backend_against_coincurve.py]: What's the point of reversing direction on this? Looks like copy-pasta, to me.

**Review by carver — APPROVED** [review]
Will wait a bit to see if Piper has more to add. Looks like one of the tests can be simplified.

File: tests/backends/test_native_backend_against_coincurve.py (hunk 4 above)
Comment (jannikluhn) [reply tests/backends/test_native_backend_against_coincurve.py]: Whoops, you're right, there's no point. Fixed.

**Comment by carver** [comment]
Anything else to add today @pipermerriam ? If not, I'll merge by EOD.

File: eth_keys/backends/coincurve.py [hunk 5]
```diff
@@ -67,3 +71,17 @@ def private_key_to_public_key(self, private_key: PrivateKey) -> PublicKey:
...
+    def decompress_public_key_bytes(self,
+                                    compressed_public_key_bytes: bytes) -> bytes:
+        public_key = self.keys.PublicKey(compressed_public_key_bytes)
+        return public_key.format(compressed=False)[1:]
+
+    def compress_public_key_bytes(self,
+                                  uncompressed_public_key_bytes: bytes) -> bytes:
+        point = (
+            big_endian_to_int(uncompressed_public_key_bytes[:32]),
+            big_endian_to_int(uncompressed_public_key_bytes[32:]),
```
Comment (pipermerriam) [inline eth_keys/backends/coincurve.py:88]: This allows for an arbitrarily long second value were the provided `uncompressed_public_key_bytes` to be longer than `64` bytes.  Am I missing something?  Seems like this would cause weird/hard-to-debug downstream errors in some cases.

File: eth_keys/backends/native/ecdsa.py [hunk 6]
```diff
@@ -60,6 +60,35 @@ def private_key_to_public_key(private_key_bytes: bytes) -> bytes:
...
+
+def decompress_public_key(compressed_public_key_bytes: bytes) -> bytes:
+    if len(compressed_public_key_bytes) != 33:
+        raise ValueError("Invalid compressed public key")
+
+    prefix = compressed_public_key_bytes[0]
+    if prefix not in (2, 3):
+        raise ValueError("Invalid compressed public key")
+
+    x = big_endian_to_int(compressed_public_key_bytes[1:])
```
Comment (pipermerriam) [inline eth_keys/backends/native/ecdsa.py:80]: I've been on a big `memoryview` kick recently so feel free to ignore me.  This would be slightly faster using a `memoryview` to prevent the copy.  This is only safe if the thing you're passing it to doesn't try to mutate it which in this case I believe that condition is satisfied.

**Review by pipermerriam — APPROVED** [review]
looks good.

The hypothesis tests are probably only running a small number of runs IIRC so you might do a couple of local runs or repeat the hypothesis test runs on a loop for a while to stress test them a bit and see if you find any edge cases that demonstrate the two implementations diverging before you click merge.

File: eth_keys/backends/coincurve.py (hunk 5 above)
Comment (jannikluhn) [reply eth_keys/backends/coincurve.py:88]: This is not part of the public API, all inputs should pass through `datatypes.PublicKey`. I still added validation now, just to be safe.

File: eth_keys/backends/native/ecdsa.py (hunk 6 above)
Comment (jannikluhn) [reply eth_keys/backends/native/ecdsa.py:80]: I did a quick benchmark and didn't see any improvement. I think the reason is that it would save one copy when slicing, but it adds one when creating the `memoryview`. But I will keep it in mind in the future (could be very useful for pyrlp/ssz).

**Comment by jannikluhn** [comment]
I don't have write access here, can someone merge please?

**Comment by carver** [comment]
Would you like a release after #58 is merged?

**Comment by jannikluhn** [comment]
Thanks!

> Would you like a release after #58 is merged?

Yep, that would be nice.
